File: Contest/ConvertSortedListToBinarySearchTree.java
// https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/
// @author: anuj0503

class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        return binarySearchTree(null, head);
    }
    
    private TreeNode binarySearchTree(ListNode last, ListNode head){
        if(last == head || head == null){
            return null;
        }
        
        if(last == head.next){
            return new TreeNode(head.val);
        }
        
        // System.out.println(last + " " + head.val);
        ListNode fast = head, slow = head;
        while(fast != last && fast.next != last){
            // System.out.println(fast.val);
            fast = fast.next.next;
            slow = slow.next;
        }
        
        TreeNode node = new TreeNode(slow.val);
        node.left = binarySearchTree(slow, head);
        node.right = binarySearchTree(last, slow.next);
        
        return node;
    }
}

File: Leetcode2023/March2023/C++/Convert Sorted List to Binary Search Tree.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if(!head) return NULL;
        ListNode* node=head;
        while(node->next) node=node->next;

        return getTreeNode(head);
    }

    TreeNode* getTreeNode(ListNode* l)
    {
        if(l==NULL) return NULL;
        if(!l->next) return new TreeNode(l->val);

        ListNode* currNode=getMiddleNode(l);
        ListNode* temp =currNode->next;
        currNode->next=NULL;
        TreeNode* curr=new TreeNode(currNode->val);
        curr->left=getTreeNode(l);
        curr->right=getTreeNode(temp);
        return curr;
    }

    ListNode* getMiddleNode(ListNode* l)
    {
        ListNode* slow= l;
        ListNode* fast = l,*prev= NULL;
        while(fast && fast->next)
        {
            prev=slow;
            slow= slow->next;
            fast=fast ? fast->next?fast->next->next:fast->next:fast;
        }
        prev->next = NULL;
        return slow;

    }

};